models/classes/resources/Service/InstanceMetadataCopier.php

```diff
@@ -24,15 +24,137 @@
+
+    /** @var string[] */
+    private $blacklistedProperties = [
+        OntologyRdf::RDF_TYPE,
+    ];
+
+    /** @var array<string, core_kernel_classes_Property> */
+    private $sharedProperties = [];
+
+    /** @var array<string, bool> */
+    private $fileProperties = [];
+
+    public function __construct(
+        ClassMetadataMapperInterface $classMetadataMapper,
+        FileReferenceSerializer $fileReferenceSerializer,
+        FileSystemService $fileSystemService
+    ) {
+        $this->classMetadataMapper = $classMetadataMapper;
+        $this->fileReferenceSerializer = $fileReferenceSerializer;
+        $this->fileSystemService = $fileSystemService;
+    }
+
+    public function addPropertyUriToBlacklist(string $propertyUri): void
+    {
+        if (!in_array($propertyUri, $this->blacklistedProperties, true)) {
+            $this->blacklistedProperties[] = $propertyUri;
+        }
+    }
+
     public function copy(
         core_kernel_classes_Resource $instance,
         core_kernel_classes_Resource $destinationInstance
     ): void {
-        // TODO: Implement copy() method.
+        $originalClass = current($instance->getTypes());
+        $destinationClass = current($destinationInstance->getTypes());
+
+        $destinationProperties = $destinationClass->getProperties(true);
+        $this->sharedProperties = array_merge(
+            $this->sharedProperties,
+            array_intersect_key($originalClass->getProperties(), $destinationProperties)
+        );
+
+        foreach ($destinationProperties as $destinationProperty) {
+            $originalProperty = $this->getOriginalProperty($destinationProperty);
+
+            if (
+                $originalProperty === null
+                || in_array($originalProperty->getUri(), $this->blacklistedProperties, true)
+            ) {
+                continue;
+            }
+
+            $propertyValues = $instance->getPropertyValuesCollection($originalProperty);
+
+            /** @var core_kernel_classes_Literal|core_kernel_classes_Resource|core_kernel_classes_Resource[] $propertyValue */
+            foreach ($propertyValues->getIterator() as $propertyValue) {
+                if ($this->isFileProperty($originalProperty)) {
+                    $this->copyFile($destinationInstance, $destinationProperty, $propertyValue);
+
+                    continue;
+                }
+
+                $destinationInstance->setPropertyValue($destinationProperty, $propertyValue);
+            }
+        }
+    }
+
+    private function getOriginalProperty(core_kernel_classes_Property $property): ?core_kernel_classes_Property
+    {
+        if (array_key_exists($property->getUri(), $this->sharedProperties)) {
+            return $property;
+        }
+
+        $originalPropertyUri = $this->classMetadataMapper->get($property);
+
+        return $originalPropertyUri === null
+            ? null
+            : $property->getProperty($originalPropertyUri);
```

Comment:
suggestion: The ternary operator here may make the code slightly harder to read, I would try to avoid it

```suggestion
        if ($originalPropertyUri === null) {
            return null;
        }
        
        return $property->getProperty($originalPropertyUri);
```